File: pets/extensions/models/apt_repository.ts
import { z } from "npm:zod@4";
import {
  errorMessage,
  type FleetContext,
  fleetFacts,
  logHostStatus,
  nowIso,
  pyScript,
  sha256Hex,
  type WriteResourceFn,
} from "./_lib/fleet.ts";
import { source as aptRepositoryPy } from "./_payloads/apt_repository.ts";
import {
  aptRepositoryApplyCfg,
  aptRepositoryApplyFacts,
  aptRepositoryCheckCfg,
  aptRepositoryCheckFacts,
} from "./_lib/cfg.ts";

const GlobalArgsSchema = z.object({
  fleet: z.string().min(1).describe(
    "@swamp/ssh model providing inventory and transport",
  ),
  name: z.string().regex(/^[a-z0-9-]+$/).describe(
    "Filename stem for /etc/apt/sources.list.d/<name>.sources",
  ),
  uris: z.array(z.string().url()).min(1).describe("Repository base URLs"),
  suites: z.array(z.string()).min(1).describe(
    "Distribution suites (e.g. trixie-pgdg)",
  ),
  components: z.array(z.string()).min(1).describe(
    "Repository components (e.g. main)",
  ),
  architectures: z.array(z.string()).optional().describe(
    "Restrict to these architectures",
  ),
  gpgKeyUrl: z.string().url().optional().describe(
    "Key fetched to signedBy when absent (.asc target: stored armored as-is)",
  ),
  signedBy: z.string().startsWith("/").optional().describe(
    "Keyring path referenced by Signed-By",
  ),
});

/** Parsed global arguments (exported for tests). */
export type GlobalArgs = z.infer<typeof GlobalArgsSchema>;

interface Ctx extends FleetContext {
  globalArgs: GlobalArgs;
  writeResource: WriteResourceFn;
}

const HostsArg = z.object({
  hosts: z.string().min(1).describe("Fleet host selector (name:X or tag:Y)"),
});

const StateSchema = z.object({
  path: z.string(),
  status: z.enum(["compliant", "non_compliant", "applied", "failed"]),
  sourcesMatch: z.boolean(),
  keyPresent: z.boolean(),
  changes: z.array(z.string()),
  error: z.string().nullable(),
  timestamp: z.string(),
});

type RepoFacts = z.infer<typeof aptRepositoryCheckFacts>;

function sourcesPath(g: GlobalArgs): string {
  return `/etc/apt/sources.list.d/${g.name}.sources`;
}

/** Render the deb822 .sources content exactly as apply installs it. */
export function deb822(g: GlobalArgs): string {
  const lines = ["Types: deb"];
  lines.push(`URIs: ${g.uris.join(" ")}`);
  lines.push(`Suites: ${g.suites.join(" ")}`);
  lines.push(`Components: ${g.components.join(" ")}`);
  if (g.architectures) {
    lines.push(`Architectures: ${g.architectures.join(" ")}`);
  }
  if (g.signedBy) lines.push(`Signed-By: ${g.signedBy}`);
  return lines.join("\n") + "\n";
}

function checkCfg(g: GlobalArgs) {
  return { path: sourcesPath(g), signedBy: g.signedBy ?? null };
}

function judge(g: GlobalArgs, expected: string, facts: RepoFacts) {
  const sourcesMatch = facts.fileHash === expected;
  const changes = [
    ...(sourcesMatch ? [] : [`write ${sourcesPath(g)}`]),
    ...(facts.keyPresent ? [] : [`fetch key to ${g.signedBy}`]),
  ];
  return { sourcesMatch, changes };
}

/**
 * Manages a deb822 `.sources` apt repository entry and its signing key.
 *
 * Renders the .sources file exactly as the official Docker Debian docs
 * prescribe (deb822, Types/URIs/Suites/Components/Architectures/Signed-By,
 * trailing newline); keys ending in .asc are installed armored, without
 * dearmoring, matching those docs.
 */
export const model = {
  type: "@psftw/pets/apt-repository",
  version: "2026.08.12.1",
  globalArguments: GlobalArgsSchema,
  inputsSchema: z.object({}),
  resources: {
    state: {
      description: "Per-host apt source compliance",
      schema: StateSchema,
      lifetime: "infinite",
      garbageCollection: 50,
    },
  },
  methods: {
    check: {
      description:
        "Compare .sources content hash and key presence (read-only, no sudo)",
      arguments: HostsArg,
      execute: async (args: { hosts: string }, context: Ctx) => {
        const g = context.globalArgs;
        const expected = await sha256Hex(deb822(g));
        const script = pyScript(
          aptRepositoryPy,
          "check",
          aptRepositoryCheckCfg,
          checkCfg(g),
        );
        const runs = await fleetFacts(
          context,
          g.fleet,
          args.hosts,
          script,
          aptRepositoryCheckFacts,
        );
        const handles = [];
        for (const run of runs) {
          const j = judge(g, expected, run.data);
          const status = j.changes.length === 0 ? "compliant" : "non_compliant";
          logHostStatus(context, run.host, status, j.changes);
          handles.push(
            await context.writeResource("state", run.host, {
              path: sourcesPath(g),
              status,
              sourcesMatch: j.sourcesMatch,
              keyPresent: run.data.keyPresent,
              changes: j.changes,
              error: null,
              timestamp: nowIso(),
            }),
          );
        }
        return { dataHandles: handles };
      },
    },
    apply: {
      description:
        "Install the .sources file and fetch the signing key if absent (sudo); no-op when compliant",
      arguments: HostsArg,
      execute: async (args: { hosts: string }, context: Ctx) => {
        const g = context.globalArgs;
        const expected = await sha256Hex(deb822(g));
        const script = pyScript(
          aptRepositoryPy,
          "apply",
          aptRepositoryApplyCfg,
          {
            ...checkCfg(g),
            contentB64: btoa(deb822(g)),
            gpgKeyUrl: g.gpgKeyUrl ?? null,
          },
        );
        try {
          const runs = await fleetFacts(
            context,
            g.fleet,
            args.hosts,
            script,
            aptRepositoryApplyFacts,
            { sudo: true, timeoutSec: 120 },
          );
          const handles = [];
          for (const run of runs) {
            const j = judge(g, expected, run.data);
            const error = run.data.error ??
              (j.changes.length ? "post-apply verification failed" : null);
            const changes = [
              ...(run.data.wroteSources ? [`write ${sourcesPath(g)}`] : []),
              ...(run.data.fetchedKey ? [`fetch key to ${g.signedBy}`] : []),
            ];
            const status = error !== null
              ? "failed"
              : changes.length === 0
              ? "compliant"
              : "applied";
            logHostStatus(context, run.host, status, changes);
            handles.push(
              await context.writeResource("state", run.host, {
                path: sourcesPath(g),
                status,
                sourcesMatch: j.sourcesMatch,
                keyPresent: run.data.keyPresent,
                changes,
                error,
                timestamp: nowIso(),
              }),
            );
          }
          return { dataHandles: handles };
        } catch (err) {
          await context.writeResource("state", args.hosts, {
            path: sourcesPath(g),
            status: "failed",
            sourcesMatch: false,
            keyPresent: false,
            changes: [],
            error: errorMessage(err),
            timestamp: nowIso(),
          });
          throw err;
        }
      },
    },
  },
  reports: [],
};

File: pets/payloads/apt_repository.py
"""Deb822 apt source + signing key facts and apply for @psftw/pets/apt-repository.

Keys ending in .asc are stored armored as-is (official Docker Debian docs).
"""
import base64
import hashlib
import json
import os
import subprocess
import tempfile
import urllib.request

# BEGIN GENERATED TYPES — edit extensions/models/_lib/cfg.ts and run `deno task gen`.
from typing import TypedDict


class CheckCfg(TypedDict):
    path: str
    signedBy: str | None


class ApplyCfg(TypedDict):
    path: str
    signedBy: str | None
    contentB64: str
    gpgKeyUrl: str | None
# END GENERATED TYPES


def gather(cfg: CheckCfg):
    out: dict[str, object] = {"fileHash": None, "keyPresent": True}
    if os.path.isfile(cfg["path"]):
        with open(cfg["path"], "rb") as f:
            out["fileHash"] = hashlib.sha256(f.read()).hexdigest()
    if cfg["signedBy"]:
        out["keyPresent"] = (
            os.path.isfile(cfg["signedBy"]) and os.path.getsize(cfg["signedBy"]) > 0
        )
    return out


def check(cfg: CheckCfg):
    print(json.dumps(gather(cfg)))


def _install(dest: str, data: bytes):
    fd, tmp = tempfile.mkstemp(dir=os.path.dirname(dest))
    try:
        with os.fdopen(fd, "wb") as f:
            f.write(data)
        os.chmod(tmp, 0o644)
        os.replace(tmp, dest)
    except BaseException:
        try:
            os.unlink(tmp)
        except FileNotFoundError:
            pass
        raise


def _fetch_key(signed: str, url: str):
    d = os.path.dirname(signed)
    os.makedirs(d, exist_ok=True)
    os.chmod(d, 0o755)
    data = urllib.request.urlopen(url, timeout=30).read()
    if not signed.endswith(".asc"):
        r = subprocess.run(["gpg", "--dearmor"], input=data, capture_output=True)
        if r.returncode != 0:
            raise RuntimeError("gpg --dearmor failed: " + r.stderr.decode()[:500])
        data = r.stdout
    _install(signed, data)


def apply(cfg: ApplyCfg):
    err = None
    wrote_sources = False
    fetched_key = False
    try:
        facts = gather(cfg)
        if cfg["signedBy"] and cfg["gpgKeyUrl"] and not facts["keyPresent"]:
            _fetch_key(cfg["signedBy"], cfg["gpgKeyUrl"])
            fetched_key = True
        content = base64.b64decode(cfg["contentB64"])
        if facts["fileHash"] != hashlib.sha256(content).hexdigest():
            _install(cfg["path"], content)
            wrote_sources = True
    except Exception as e:
        err = str(e)[:2000]
    out = gather(cfg)
    out["wroteSources"] = wrote_sources
    out["fetchedKey"] = fetched_key
    out["error"] = err
    print(json.dumps(out))
